### PR TITLE
Fix UB/crash in new SeqnoToTimeMapping::CopyFromSeqnoRange

### PR DESCRIPTION
**Pull request by pdillinger**
Summary: After #12253 this function has crashed in the crash test, in its call to `std::copy`. I haven't reproduced the crash directly, but `std::copy` probably has undefined behavior if the starting iterator is after the ending iterator, which was possible. I've fixed the logic to deal with that case and to add an assertion to check that precondition of `std::copy` (which appears can be unchecked by `std::copy` itself even with UBSAN+ASAN).

Also added some unit tests etc. that were unfinished for #12253, and slightly tweak SeqnoToTimeMapping::EnforceMaxTimeSpan handling of zero time span case.

This is intended for patching 8.11.

Test Plan: tests added. Will trigger ~20 runs of the crash test job that saw the crash.

### CODE REVIEW

**Review by jowlyzhang — APPROVED** [review]
LGTM, thank you for the improvement!

Just for my learning, why is it preferable to make 0 timespan also have one entry, instead of keeping it empty. Is it to be more logically consistent with the keeping one entry before cutoff invariant?

File: db/seqno_to_time_mapping.cc [hunk 1]
```diff
@@ -400,12 +395,16 @@ void SeqnoToTimeMapping::CopyFromSeqnoRange(const SeqnoToTimeMapping& src,
                                             SequenceNumber to_seqno) {
   bool orig_empty = Empty();
   auto src_it = src.FindGreaterEqSeqno(from_seqno);
+  // Allow nonsensical ranges like [1000, 0] which might show up e.g. for
+  // an SST file with no entries.
+  auto src_it_end =
+      to_seqno < from_seqno ? src_it : src.FindGreaterSeqno(to_seqno);
```
[the remark this answers is not in the document]
Comment (pdillinger) [reply db/seqno_to_time_mapping.cc:401]: If ==, they could find different iterators (FindGreaterEqSeqno vs. FindGreaterSeqno)

**Comment by pdillinger** [comment]
> Just for my learning, why is it preferable to make 0 timespan also have one entry, instead of keeping it empty. Is it to be more logically consistent with the keeping one entry before cutoff invariant?

Yes, consistent with keeping one entry <= the cutoff time :)

File: db/seqno_to_time_mapping.cc (hunk 1 above)
[the remark this answers is not in the document]
Comment (jowlyzhang) [reply db/seqno_to_time_mapping.cc:401]: Ah, I missed that. Thanks for the explanation.

**Pull request closed without merging**